toolchain/sem_ir/inst_fingerprinter.cpp

```diff
@@ -0,0 +1,338 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/sem_ir/inst_fingerprinter.h"
+
+#include "common/ostream.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StableHashing.h"
+#include "toolchain/base/value_ids.h"
+#include "toolchain/sem_ir/ids.h"
+#include "toolchain/sem_ir/typed_insts.h"
+
+namespace Carbon::SemIR {
+
+namespace {
+struct Worklist {
+  // The file containing the instruction we're currently processing.
+  const File* sem_ir;
+  // The instructions we need to compute fingerprints for.
+  llvm::SmallVector<std::pair<const File*, InstId>> todo;
+  // The contents of the current instruction as accumulated so far. This is used
+  // to build a Merkle tree containing a fingerprint for the current
+  // instruction.
+  llvm::SmallVector<llvm::stable_hash> contents = {};
+  // Known cached instruction fingerprints.
+  Map<std::pair<const File*, InstId>, uint64_t>* fingerprints;
+
+  // Add an invalid marker to the contents. This is used when the entity
+  // contains an invalid ID.
+  auto AddInvalid() -> void { contents.push_back(-1); }
+
+  // Add a string to the contents.
+  auto AddString(llvm::StringRef string) {
+    contents.push_back(llvm::stable_hash_name(string));
+  }
+
+  // Each of the following `Add` functions adds a typed argument to the contents
+  // of the current instruction. If we don't yet have a fingerprint for the
+  // argument, it instead adds that argument to the worklist instead.
+
+  auto Add(InstKind kind) -> void {
+    // TODO: Precompute or cache the hash of instruction IR names, or pick a
+    // scheme that doesn't change when IR names change.
+    AddString(kind.ir_name());
+  }
+
+  auto Add(IdentifierId ident_id) -> void {
+    AddString(ident_id.is_valid() ? sem_ir->identifiers().Get(ident_id) : "");
```

Comment:
Why not AddInvalid() if not valid here? If there's a reason maybe a comment could explain?